### PR TITLE
Added Quicksort [Javascript]

### LINKED ISSUE #405: Quicksort [Javascript]
#### This issue is about [Quicksort](https://en.wikipedia.org/wiki/Quicksort).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by rashinanwani**
Fixes #405

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

File: quicksort/quickSort.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/*
+*	Implementation of quicksort algorithm
+*	Time complexity: O(nlogn)
+*	Space complexity: O(n)
+*/
+function partition (arr, start, end) {
+	let tmp;
+	for (let i = start; i < end; i++) {
+		if (arr[i] <= arr[end]) {
```
Comment (Monal5031) [inline quicksort/quickSort.js]: Please randomize the pivot. See python implementation for how  to randomize the pivot.

File: quicksort/quickSort.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+			arr[start - 1] = tmp;
+		}
+	}
+	tmp = arr[start];
+	arr[start] = arr[end];
+	arr[end] = tmp;
+	return start;
+}
+
+function quickSort (arr, left, right) {
```
Comment (Monal5031) [inline quicksort/quickSort.js]: Make left and right as default parameters with values `0` and `arr.length-1`

File: quicksort/quickSort.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		let p = partition(arr, left, right);
+		quickSort(arr, left, p - 1);
+		quickSort(arr, p + 1, right);
+	}
+}
+
+function main () {
+	let arr = [2, 1, 6, 444, 8, 99, 10];
+	console.log('Sorted data is');
+	quickSort(arr, 0, arr.length - 1);
```
Comment (Monal5031) [inline quicksort/quickSort.js]: No need to mentioned the extra parameters now.

File: quicksort/quickSort.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/*
+*	Implementation of quicksort algorithm
+*	Time complexity: O(nlogn)
+*	Space complexity: O(n)
+*/
+function partition (arr, start, end) {
+	let pivot = arr[end];
```
Comment (Monal5031) [inline quicksort/quickSort.js]: Please randomize pivot, as its done in python implementation. You didn't randomize it.

**Comment by Monal5031** [comment]
@rashinanwani after making changes, ask anyone from @iiitv/algos-support to mark the PR as `pending review`.

File: quicksort/quickSort.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+		quickSort(arr, left, p - 1);
+		quickSort(arr, p + 1, right);
+	}
+}
+
+function main () {
+	let arr = [2, 1, 6, 444, 8, 99, 10];
+	console.log('Sorted data is');
+	quickSort(arr, 0, arr.length - 1);
+	for (let i = 0; i < arr.length; i++) {
```
Comment (Monal5031) [inline quicksort/quickSort.js]: You can directly log the sorted array, no need for loop here.

File: quicksort/quickSort.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+			arr[pi] = tmp;
+		}
+	}
+	tmp = arr[pi + 1];
+	arr[pi + 1] = arr[end];
+	arr[end] = tmp;
+	return pi + 1;
+}
+
+function quickSort (arr, left = 0, right = arr.length - 1) {
```
Comment (Monal5031) [inline quicksort/quickSort.js]: Add comments about parameters.

**Comment by Monal5031** [comment]
@rashinanwani Please resolve conflicts.

File: quicksort/quickSort.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	if (arr[i] > arr[j] && arr[i] > arr[j]) {
+		return i;
+	} else if (arr[j] > arr[i] && arr[j] > arr[k]) {
+		return j;
+	} else {
+		return k;
+	}
+}
+
+function partition (arr, start, end) {
```
Comment (Monal5031) [inline quicksort/quickSort.js]: Docstrings?

File: quick_sort/quickSort.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+* 	Space complexity: O(n)
+*/
+function median (arr, i, j, k) {
+	/*
+	:param arr: array of elements
+	:param i: index of first element
+	:param j: index of last element
+	:param k: index of middle element
+	*/
+	if (arr[i] > arr[j] && arr[i] > arr[j]) {
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: 😲 😨

File: quick_sort/quickSort.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		return k;
+	}
+}
+
+function partition (arr, start, end) {
+	/*
+	:param arr: array of elements
+	:param start: index of first element
+	:param end: index of last element
+	*/
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: return?

File: quick_sort/quickSort.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+*	Time complexity: O(nlogn)
+* 	Space complexity: O(n)
+*/
+function median (arr, i, j, k) {
+	/*
+	:param arr: array of elements
+	:param i: index of first element
+	:param j: index of last element
+	:param k: index of middle element
+	*/
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: return?

File: quick_sort/quickSort.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
+/*
+*	Implementation of quicksort algorithm
+*	Time complexity: O(nlogn)
+* 	Space complexity: O(n)
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: Specify what is `n`, on next line

File: quick_sort/quickSort.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+*/
+function median (arr, i, j, k) {
+	/*
+	:param arr: array of elements
+	:param i: index of first element
+	:param j: index of last element
+	:param k: index of middle element
+	:return: return median of values at indices i, j and k
+	*/
+	if (arr[i] > arr[j] && arr[i] > arr[k]) {
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: Correct both the conditions, they are wrong :worried: :scream:

File: quick_sort/quickSort.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+		return i;
+	} else if (arr[j] > arr[i] && arr[j] > arr[k]) {
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: Same here, its also wrong I think

File: quick_sort/quickSort.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+function median (arr, i, j, k) {
+	/*
+	:param arr: array of elements
+	:param i: index of first element
+	:param j: index of last element
+	:param k: index of middle element
+	:return: return median of values at indices i, j and k
+	*/
+	if (arr[i] > arr[j]) {
+		if (arr[k] > arr[j]) {
```
Comment (Monal5031) [inline quick_sort/quickSort.js]: The conditions here look wrong. (also in python implementation) as there can be a case when arr[k]>arr[i]>arr[j], in this case median is supposed to be `arr[i]` but we get k instead.

File: quick_sort/quickSort.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+		quickSort(arr, left, p - 1);
+		quickSort(arr, p + 1, right);
+	}
+}
+
+function main () {
+	let arr = [2, 1, 6, 44, 8, 9, 10];
+	console.log('Sorted data is');
+	quickSort(arr);
+	console.log(arr);
```
Comment (Anshuman-Verma) [inline quick_sort/quickSort.js]: I don't think we need multiple `console.log()`, using `console.log('Sorted data is', arr);` should suffice.

Comment (Anshuman-Verma) [inline quick_sort/quickSort.js]: Or perhaps using console.log(`Sorted data is ${arr}`) should work well too.

File: quick_sort/quickSort.js [hunk 16]
```diff
@@ -76,9 +76,8 @@ function quickSort (arr, left = 0, right = arr.length - 1) {
 
 function main () {
 	let arr = [2, 1, 6, 44, 8, 9, 10];
-	console.log('Sorted data is');
 	quickSort(arr);
-	console.log(arr);
+	console.log('Sorted data is',arr);
```
Comment (Anshuman-Verma) [inline quick_sort/quickSort.js]: Space after comma please

**Pull request merged**